File: python/task1.py
from collections import defaultdict
from itertools import combinations
from pyspark import SparkContext
import json
import random
import sys
import time


class T1:
    def __init__(self) -> None:
        self.ipf = sys.argv[1]
        self.opf = sys.argv[2]
        self.sc = SparkContext.getOrCreate()
        self.sc.setLogLevel("OFF")

    @staticmethod
    def gen_hash_fns(num_funcs):
        step = 10 ** 7
        curstep = step
        start = 10 ** 9 - 10 ** 8 + 17
        fns = [None] * num_funcs

        for i in range(num_funcs):
            a = random.randint(start, start + curstep - 10 ** 7 + 23)
            b = random.randint(10 ** 9 - 10 ** 8 + 30697,
                               10 ** 9 + curstep + 10 ** 8 - 4123892)
            curstep += step
            fns[i] = (a, b)
        return fns

    @staticmethod
    def gen_signatures(uid_values, hash_params, num_buckets):
        p = 10 ** 9 + 7
        sig = []
        for a, b in hash_params:
            sig_row = []
            for v in uid_values:
                sig_row.append(((a * v + b) % p) % num_buckets)
            sig.append(min(sig_row))
        return sig

    def write_data(self, actual_pairs):
        with open(self.opf, "w+") as f:
            for b1, b2, sim in actual_pairs:
                f.write(json.dumps({'b1': b1, 'b2': b2, 'sim': sim}) + "\n")

    @staticmethod
    def split_bands(whole_array, num_bands):
        ret = []
        cur_band_num = 0
        for i in range(0, len(whole_array), num_bands):
            ret.append((cur_band_num, hash(tuple(whole_array[i: i + num_bands]))))
            cur_band_num += 1
        return ret

    @staticmethod
    def sb2(A, nb):
        ret = []
        cur_band_num = 0
        for i in range(0, len(A), nb):
            ret.append(hash(tuple(A[i: i + nb])))
            cur_band_num += 1
        return ret

    def run(self):
        num_bands = 100
        num_hashes = 100
        textRDD = self.sc.textFile(self.ipf).map(lambda row: json.loads(row))
        textRDD.cache()

        userRDD = textRDD.map(lambda row: row["user_id"]).distinct()
        cmp_map = userRDD.zipWithIndex().collectAsMap()
        num_buckets = len(cmp_map) - 1
        hash_params = T1.gen_hash_fns(num_hashes)

        # (biz_id, {uid1, uid2, uid3, uid17, uid50, ...})
        biz_sets = textRDD.map(lambda row: (row['business_id'], cmp_map[row['user_id']])).distinct(
        ).groupByKey().map(lambda row: (row[0], set(row[1])))
        biz_sets.cache()
        biz_map = biz_sets.collectAsMap()

        sig_temp = biz_sets.mapValues(lambda uids_list: T1.gen_signatures(uids_list, hash_params, num_buckets)).collect()
        # print(sig_temp[:10])
        candidate_pairs = set()
        for i in range(num_bands):
            cur_bucket = defaultdict(set)
            for row in sig_temp:
                cur_bucket[row[1][i]].add(row[0])
            for v in cur_bucket.values():
                if len(v) == 1:
                    continue
                for b1, b2 in combinations(sorted(v), 2):                    
                    candidate_pairs.add((b1, b2))
            # print("Processed band: {}".format(i))
        

        # for band_num in range(num_bands):
        #     current_bucket = defaultdict(set)
        #     for sig in chunks[band_num]:
        #         # print("Sig for band: {}, {}".format(sig, band_num))
        #         hash_val = hash(tuple(sig[1]))
        #         current_bucket[hash_val].add(sig[0])
        #     # print("Processing band: {}".format(band_num))
        #     # print("Number of hash buckets: {}".format(len(current_bucket)))
        #     for v in current_bucket.values():
        #         # print("Len of current hashvalues: {}".format(len(v)))
        #         if len(v) == 1:
        #             continue
        #         for b1, b2 in combinations(sorted(current_bucket), 2):
        #             print("Potential candidates: ", b1, b2)
        #             candidate_pairs.add((b1, b2))

        actual_similar_bizz = []
        print("Number of candidate pairs: {}".format(len(candidate_pairs)))
        for b1, b2 in candidate_pairs:
            u1 = biz_map[b1]
            u2 = biz_map[b2]

            sim = len(u1 & u2) / len(u1 | u2)
            if sim >= 0.05:
                actual_similar_bizz.append((b1, b2, sim))
        self.write_data(actual_similar_bizz)
        return
        
        
        
        cands1 = sig_temp.flatMap(lambda item: [(tuple(x), item[0]) for x in T1.split_bands(item[1], num_bands)]).groupByKey().map(lambda item: list(item[1]))
        
        print("~~~ Cands1 ~~~")
        i = 0
        for k in cands1.collect():
            if len(k) > 1:
                print(k)
            

        cands2 = cands1.filter(lambda count: len(count) > 1)
        cands3 = cands2.collect()
        candidate_pairs = cands3
        
        
        # flatMap(lambda bizz: [bizz2 for bizz2 in combinations(bizz, 2)]).collect()
        # print("Reached sig_m generated")
        # print("Number of rows in sig_m: {}".format(len(sig_m)))
        # candidate_pairs = set()
        # chunks = T1.split_bands(sig_m, num_bands)
        # print("First chunk items: {}".format(chunks[0][:10]))

        


if __name__ == "__main__":
    t1 = T1()

    st = time.time()
    t1.run()
    et = time.time()
    print("Total time taken: {}".format(et - st))

File: scala/task1.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD
import org.json4s._
import org.json4s.jackson.JsonMethods._

import scala.util.control.Breaks._
import scala.util.Random._
import collection.mutable.{ArrayBuffer, HashMap, HashSet}
import java.io.{File, PrintWriter}
import scala.io.Source
import scala.math.Ordering.Implicits._
import scala.util.Random


object task1 {
  implicit val formats = DefaultFormats

  def genHashFns(numFuncs: Int): Array[(Long, Long)] = {
    val start = (10e7 - 5128423).toInt
    val end = (10e8 + 3267231).toInt
    val bstart = (10e7 - 3069127).toInt
    val bend = (10e8 + 1278127).toInt
    var fns = new ArrayBuffer[(Long, Long)]
    val random = new Random()

    for (i <- 0 to numFuncs) {
      val a = start + random.nextInt(end - start)
      val b = bstart + random.nextInt(bend - bstart)
      fns += Tuple2(a, b)
    }

    return fns.toArray
  }

  def genSignatures(uidValues: Set[Long], hashParams: Array[(Long, Long)], numBuckets: Int): Array[Long] = {
    val p = (10e9 + 7).toInt
    var sig = new ArrayBuffer[Long]

    for (params <- hashParams) {
      val a = params._1
      val b = params._2
      var sig_row = new ArrayBuffer[Long]

      for (v <- uidValues) {
        sig_row += (((a * v + b) % p) % numBuckets)
      }

      sig += sig_row.min
    }

    return sig.toArray
  }

  def writeData(simBizz: Array[(String, String, Double)], outputfile: String): Unit = {
    val out = new PrintWriter(new File(outputfile))

    for (item <- simBizz) {
      val output: Map[String, Any] = Map("b1"->item._1, "b2"->item._2, "sim"->item._3)
      val formatted_output = org.json4s.jackson.Serialization.write(output)
      out.write(formatted_output)
      out.write("\n")
    }

    out.close()
  }

  def main(args: Array[String]) = {
    val t1 = System.nanoTime

    val ipf = args(0)
    val opf = args(1)

    val sc = SparkContext.getOrCreate()
    sc.setLogLevel("OFF")

    val NUM_BANDS = 100
    val NUM_HASHES = 100

    val textRDD = sc.textFile(ipf).map(row => parse(row))
    textRDD.cache()

    val userRDD = textRDD.map(row => (row \ "user_id").extract[String]).distinct()
    val cmp_map = userRDD.zipWithIndex().collectAsMap()
    val num_buckets = cmp_map.size - 1
    val hash_params = genHashFns(NUM_HASHES)

    println("Done generating hash functions")

    val bizSets = textRDD.map(row => ((row \ "business_id").extract[String], cmp_map((row \ "user_id").extract[String])))
      .distinct()
      .groupByKey()
      .map(row => (row._1, row._2.toSet))
    bizSets.cache()
    val bizMap = bizSets.collectAsMap()

    println("Done generating bizSets and bizMap")

    val sigTemp = bizSets.mapValues(uidsList => genSignatures(uidsList, hash_params, num_buckets)).collect()

    println("Done generating sigTemp")

    var cands = new HashSet[(String, String)]()

    for (i <- 0 to NUM_BANDS - 1) {
      var curBucket = new HashMap[Long, HashSet[String]].withDefaultValue(HashSet[String]())
      for (row <- sigTemp) {
        if (curBucket.contains(row._2(i))) {
          curBucket(row._2(i)) += row._1
        } else {
          curBucket(row._2(i)) = new HashSet[String]
          curBucket(row._2(i)) += row._1
        }
      }
      for (v <- curBucket.values) {
        if (v.size > 1) {
          for (comb <- v.toArray.sorted.combinations(2)) {
            cands += Tuple2(comb(0), comb(1))
          }
        }
      }
    }

    println("Done generating Candidates")
    println(cands.size)

    var actualSimBizz = new ArrayBuffer[(String, String, Double)]
    for (pair <- cands) {
      val b1 = pair._1
      val b2 = pair._2

      val u1 = bizMap(b1)
      val u2 = bizMap(b2)

      val sim = (u1.intersect(u2)).size.toDouble / (u1.union(u2)).size.toDouble
      if (sim >= 0.05) {
        actualSimBizz += Tuple3(b1, b2, sim)
      }
    }

    println("Done generating similar pairs")

    writeData(actualSimBizz.toArray, outputfile = opf)

    println("Done writing data")
    println("Duration: " + (System.nanoTime - t1) / 1e9d)
  }
}
